swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -47,295 +72,370 @@ public class Adapter {
-      networkPrefixLengths: [networkPrefixLength])
-    ipv6Settings.includedRoutes = ipv6Routes
-    networkSettings.ipv6Settings = ipv6Settings
-
-    return networkSettings
-  }
-
-  public func setNetworkSettings(_ networkSettings: NEPacketTunnelNetworkSettings) throws {
-    var systemError: Error?
-    let condition = NSCondition()
-
-    // Activate the condition
-    condition.lock()
-    defer { condition.unlock() }
-
-    self.packetTunnelProvider?.setTunnelNetworkSettings(networkSettings) { error in
-      systemError = error
-      condition.signal()
-    }
-
-    // Packet tunnel's `setTunnelNetworkSettings` times out in certain
-    // scenarios & never calls the given callback.
-    let setTunnelNetworkSettingsTimeout: TimeInterval = 5  // seconds
-
-    if condition.wait(until: Date().addingTimeInterval(setTunnelNetworkSettingsTimeout)) {
-      if let systemError = systemError {
-        throw AdapterError.setNetworkSettings(systemError)
-      }
-    }
+// MARK: Responding to path updates
 
-    // Save the latest applied network settings if there was no error.
-    if systemError != nil {
-      self.lastNetworkSettings = networkSettings
+extension Adapter {
+  private func beginPathMonitoring() {
+    self.logger.debug("Beginning path monitoring")
+    let networkMonitor = NWPathMonitor()
+    networkMonitor.pathUpdateHandler = { [weak self] path in
+      self?.didReceivePathUpdate(path: path)
     }
+    networkMonitor.start(queue: self.workQueue)
   }
 
-  /// Update runtime configuration.
-  /// - Parameters:
-  ///   - ipv4Routes: IPv4 routes to send through the tunnel.
-  ///   - ipv6Routes: IPv6 routes to send through the tunnel.
-  ///   - completionHandler: completion handler.
-  public func update(
-    ipv4Routes: [NEIPv4Route], ipv6Routes: [NEIPv6Route],
-    completionHandler: @escaping (AdapterError?) -> Void
-  ) {
-    workQueue.async {
-      if case .stopped = self.state {
-        completionHandler(.invalidState)
-        return
-      }
-
-      // Tell the system that the tunnel is going to reconnect using new WireGuard
-      // configuration.
-      // This will broadcast the `NEVPNStatusDidChange` notification to the GUI process.
-      self.packetTunnelProvider?.reasserting = true
-      defer {
-        self.packetTunnelProvider?.reasserting = false
-      }
-
-      do {
-        try self.setNetworkSettings(
-          self.generateNetworkSettings(ipv4Routes: ipv4Routes, ipv6Routes: ipv6Routes))
-
-        switch self.state {
-        case .started(let wrappedSession):
-          self.state = .started(wrappedSession)
-
-        case .temporaryShutdown:
-          self.state = .temporaryShutdown
+  private func didReceivePathUpdate(path: Network.NWPath) {
+    // Will be invoked in the workQueue by the path monitor
+    switch self.state {
 
-        case .stopped:
-          fatalError()
+      case .startingTunnel(let session, let onStarted):
+        if path.status == .satisfied {
```

Comment:
I don't know elixir but isn't it better to do just `if path.status != .satisfied` without the `else`?

Reply:
True. Fixed in bf84476.